### PR TITLE
feat(dashboard): add raw JSON body mode to HTTP request step

### PR DESCRIPTION
**Pull request by FredRMonizze**
## Summary

The HTTP Request action step currently only supports flat key-value pairs for the request body, making it impossible to send nested JSON objects required by many third-party APIs (e.g. CM Telecom Voice API which expects a `voice: { language, gender, ... }` object).

This PR adds a toggle between **Key-Value** and **Raw JSON** modes in the HTTP Request step, allowing users to paste/type a full JSON body with nested objects, arrays, and LiquidJS variables.

Related issue: #8784

## Why

When integrating with APIs that expect nested JSON bodies, users currently have no way to send something like:

\`\`\`json
{
  "callee": "+32...",
  "voice": {
    "language": "fr",
    "gender": "Male"
  }
}
\`\`\`

Workarounds tried (all failed):
- \`voice.language\` as flat key → sent as literal \`\"voice.language\"\` key
- \`voice[language]\` notation → same issue
- Pasting JSON as a single value → string-escaped by Novu

### Schema & Types
- Added \`bodyMode\` (\`'key-value' | 'raw'\`, default \`'key-value'\`) and \`rawBody\` fields to the HTTP request control schema
- Added \`DESTINATION_BODY_MODE\` and \`DESTINATION_RAW_BODY\` to \`UiComponentEnum\`
- Updated \`HttpRequestControlDto\` with the new optional fields

### Frontend (Dashboard)
- New \`RawBodyEditor\` component with multiline textarea and LiquidJS variable support
- Toggle between **Key-Value** and **Raw JSON** modes in the HTTP Request editor
- Live JSON validation with inline error message (skipped when value contains \`{{...}}\` variables)
- Confirmation prompt when switching back to Key-Value mode with existing raw JSON content
- Updated \`CurlDisplay\` and \`buildRawCurlString\` to render raw body in the curl preview

### Backend
- New \`parseRawBody()\` utility that parses and validates raw JSON strings
- Updated \`ExecuteHttpRequestStep\` usecase to use \`parseRawBody\` when \`bodyMode === 'raw'\`
- Updated \`TestHttpEndpoint\` usecase to support raw body mode in the test feature
- Updated \`PreviewUsecase\` for HMAC signature preview consistency

### Tests
- Added unit tests for \`parseRawBody\`, \`toBodyRecord\`, and \`toHeadersRecord\`

## Backward compatibility

- \`bodyMode\` defaults to \`'key-value'\` → existing workflows are unaffected
- \`rawBody\` is optional → no migration needed
- Field validation skips raw body when in key-value mode and vice versa

## Test plan

- [x] Existing key-value mode still works as before
- [x] Switching to Raw JSON shows a multiline editor
- [x] Nested JSON objects are sent correctly to the target API (verified with CM Telecom Voice API)
- [x] LiquidJS variables (\`{{payload.x}}\`, \`{{subscriber.x}}\`) are resolved in raw body
- [x] Invalid JSON shows an inline error and prevents form save
- [x] Curl preview reflects the raw body content
- [x] Test endpoint works in raw mode
- [x] Switching back to Key-Value mode prompts for confirmation if raw body is non-empty
- [x] Unit tests for \`parseRawBody\` pass

---

> [!NOTE]
> **Medium Risk**
> Adds a new raw JSON body path that is parsed/validated and used for request signing/sending; invalid JSON or unexpected body types can cause runtime failures in HTTP request execution/testing/preview. Defaults preserve existing behavior, but this touches both dashboard UX and worker/api execution paths.
> 
> **Overview**
> Adds support for sending HTTP Request step bodies as **raw JSON** in addition to the existing key-value pairs.
> 
> Updates shared schema/DTOs to include `bodyMode` and `rawBody`, adds a dashboard `RawBodyEditor` with JSON validation and a mode toggle (with confirmation when switching back), and updates cURL previews/copy to reflect raw bodies.
> 
> On the backend/worker, introduces `parseRawBody` and wires it into HTTP request execution, test endpoint, and preview signature generation so raw JSON is parsed as an object before signing/sending; includes unit tests for the new parsing utility.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 15f7166daaf173e417941a45970fbc6efde4ae5e. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## What changed

The PR adds support for raw JSON request bodies in HTTP request workflow steps, complementing the existing key-value pair mode. Users can now toggle between "key-value" and "raw" modes to send nested objects, arrays, and LiquidJS-templated JSON. The change includes a new frontend editor component with live JSON validation, backend parsing and validation of raw body content, curl preview updates, and signature generation for both modes. Backward compatibility is preserved with "key-value" as the default mode.

## Affected areas

**dashboard**: Added `RawBodyEditor` component with client-side JSON validation and error messaging; updated `HttpRequestEditor` to toggle between key-value and raw modes with confirmation when switching back to key-value with unsaved content. Modified `CurlDisplay` and `curl-utils` to render raw JSON bodies directly and added shell escaping for safe curl output.

**api**: Updated `TestHttpEndpointUsecase` to parse and validate raw JSON body, returning a 400 error with descriptive message on invalid JSON. Extended `ResolvedRequestDto` to support array payloads alongside objects. Added E2E test for the test-http-endpoint endpoint covering both canonical (raw) and legacy (key-value) body formats.

**worker**: Modified `ExecuteHttpRequestStep` usecase to conditionally parse raw body via new `parseRawBody()` utility when `bodyMode === 'raw'`, with error handling that prevents request execution and returns a failed status.

**shared**: Extended HTTP request control schema and UI configuration to include `bodyMode` enum and `rawBody` string fields with appropriate UI components.

**application-generic**: Added `parseRawBody()` utility that validates JSON is a non-null object or array; updated `shouldIncludeBody()` to accept array payloads; modified `PreviewUsecase` to derive signature payload from raw body when in raw mode. Extended `HttpRequestControlDto` with `bodyMode` and `rawBody` validation.

## Key technical decisions

- Raw JSON validation rejects arrays at the root level (only objects allowed) to maintain consistency with key-value mode semantics.
- Validation is skipped if the input contains `{{...}}` template syntax to allow LiquidJS variables.
- Mode toggle from raw to key-value prompts user confirmation if unsaved raw content exists, preventing accidental data loss.
- Shell escaping applied to all curl arguments (URL, headers, body, signature) for safe POSIX compliance.

## Testing

Unit tests added for `parseRawBody`, `toBodyRecord`, and `toHeadersRecord` utilities. E2E test added for the test-http-endpoint endpoint verifying both raw and key-value body handling and signature generation. Manual testing recommended for edge cases with large payloads and complex LiquidJS templates.

### CODE REVIEW

File: apps/dashboard/src/components/workflow-editor/steps/http-request/curl-display.tsx [hunk 1]
```diff
@@ -8,27 +8,39 @@ type CurlDisplayProps = {
...
+        if (pairs.length > 0) {
+          bodyObj = Object.fromEntries(pairs.map(({ key, value }) => [key, value]));
+        }
+      } else if (Object.keys(body).length > 0) {
+        bodyObj = body;
+      }
+
+      if (bodyObj) {
+        bodyStr = JSON.stringify(bodyObj);
       }
```
[the remark this answers is not in the document]
Comment (SuperFredR) [reply apps/dashboard/src/components/workflow-editor/steps/http-request/curl-display.tsx]: Fixed in 4b6e451: `CurlDisplay` now uses an exclusive raw mode — when `bodyMode === 'raw'`, it only renders `rawBody` (or nothing if empty), never falling back to the key-value `body`.

File: apps/dashboard/src/components/workflow-editor/steps/http-request/curl-utils.ts [hunk 2]
```diff
@@ -28,21 +30,29 @@ export function buildRawCurlString(
...
+      } else if (body && Object.keys(body).length > 0) {
+        bodyObj = body;
+      }
+
+      if (bodyObj) {
+        bodyStr = `--data '${JSON.stringify(bodyObj)}'`;
       }
-    } else if (body && Object.keys(body).length > 0) {
-      bodyObj = body;
     }
```
[the remark this answers is not in the document]
Comment (SuperFredR) [reply apps/dashboard/src/components/workflow-editor/steps/http-request/curl-utils.ts:127]: Fixed in 4b6e451: `buildRawCurlString` now uses exclusive raw mode — no fallback to key-value pairs when `bodyMode === 'raw'`.

File: apps/dashboard/src/components/workflow-editor/steps/http-request/curl-utils.ts [hunk 3]
```diff
@@ -28,21 +30,29 @@ export function buildRawCurlString(
...
 
   const canHaveBody = canMethodHaveBody(method);
-  let bodyObj: Record<string, unknown> | null = null;
+  let bodyStr = '';
 
   if (canHaveBody) {
-    if (Array.isArray(body)) {
-      const pairs = body.filter((b) => b.key);
+    if (bodyMode === 'raw' && rawBody) {
+      bodyStr = `--data '${rawBody}'`;
```
[the remark this answers is not in the document]
Comment (SuperFredR) [reply apps/dashboard/src/components/workflow-editor/steps/http-request/curl-utils.ts]: Addressed in 4b6e451 along with the other curl-utils raw mode fix.

File: apps/dashboard/src/components/workflow-editor/steps/http-request/http-request-editor.tsx [hunk 4]
```diff
@@ -27,6 +31,20 @@ export function HttpRequestEditor({ uiSchema }: HttpRequestEditorProps) {
...
+  const handleBodyModeChange = (mode: 'key-value' | 'raw') => {
+    if (mode === 'key-value' && bodyMode === 'raw' && rawBody?.trim()) {
+      const confirmed = window.confirm(
+        'Switching to Key-Value mode will discard your raw JSON body. Continue?'
+      );
+      if (!confirmed) return;
+    }
+
+    setValue('bodyMode', mode);
+    saveForm();
```
[the remark this answers is not in the document]
Comment (SuperFredR) [reply apps/dashboard/src/components/workflow-editor/steps/http-request/http-request-editor.tsx]: Fixed in 098c628: `handleBodyModeChange` now also calls `setValue('rawBody', '')` after the user confirms the discard, so the state matches the message.

File: apps/dashboard/src/components/workflow-editor/steps/http-request/http-request-editor.tsx [hunk 5]
```diff
@@ -39,11 +57,46 @@ export function HttpRequestEditor({ uiSchema }: HttpRequestEditorProps) {
...
+                className={`cursor-pointer rounded px-2 py-0.5 text-xs font-medium ${
+                  bodyMode === 'raw'
+                    ? 'bg-neutral-alpha-200 text-text-strong'
+                    : 'text-text-sub hover:text-text-strong'
+                }`}
+                onClick={() => handleBodyModeChange('raw')}
+                onKeyDown={(e) => e.key === 'Enter' && handleBodyModeChange('raw')}
+              >
+                Raw JSON
+              </span>
```
[the remark this answers is not in the document]
Comment (SuperFredR) [reply apps/dashboard/src/components/workflow-editor/steps/http-request/http-request-editor.tsx]: Fixed in 4b6e451: replaced the custom span buttons with the `ToggleGroup` / `ToggleGroupItem` primitives. ARIA, keyboard handling, and focus states are now native.

File: apps/worker/src/app/workflow/usecases/send-message/execute-http-request-step.usecase.ts [hunk 6]
```diff
@@ -144,7 +147,7 @@ export class ExecuteHttpRequestStep extends SendMessageType {
     }
 
     const headersRecord = toHeadersRecord(rawHeaders);
-    const bodyObject = toBodyRecord(rawBody);
+    const bodyObject = bodyMode === 'raw' && rawJsonBody ? parseRawBody(rawJsonBody) : toBodyRecord(bodyPairs);
```
[the remark this answers is not in the document]
Comment (SuperFredR) [reply apps/worker/src/app/workflow/usecases/send-message/execute-http-request-step.usecase.ts]: Fixed in 4b6e451: `parseRawBody` is now wrapped in try/catch in `execute-http-request-step.usecase.ts`. On parse failure, an `ExecutionDetails` record is created with the error message and the step returns FAILED status (respecting `continueOnFailure`).

File: libs/application-generic/src/dtos/workflow/controls/http-request-control.dto.ts [hunk 7]
```diff
@@ -48,6 +48,21 @@ export class HttpRequestControlDto {
   @IsOptional()
   body?: HttpRequestKeyValuePairDto[];
 
+  @ApiPropertyOptional({
+    description: 'Body input mode: key-value pairs or raw JSON',
+    enum: ['key-value', 'raw'],
+  })
+  @IsString()
+  @IsOptional()
+  bodyMode?: 'key-value' | 'raw';
```
[the remark this answers is not in the document]
Comment (SuperFredR) [reply libs/application-generic/src/dtos/workflow/controls/http-request-control.dto.ts]: Fixed in 4b6e451: added `@IsIn(['key-value', 'raw'])` decorator on the `bodyMode` field.

File: libs/application-generic/src/usecases/preview/preview.usecase.ts [hunk 8]
```diff
@@ -212,9 +212,12 @@ export class PreviewUsecase {
...
       );
 
-      const rawBody = resolvedOutputs?.body as Array<{ key: string; value: string }> | undefined;
+      const bodyPairs = resolvedOutputs?.body as Array<{ key: string; value: string }> | undefined;
+      const bodyMode = (resolvedOutputs?.bodyMode as string) ?? 'key-value';
+      const rawJsonBody = resolvedOutputs?.rawBody as string | undefined;
       const method = (resolvedOutputs?.method as string) ?? 'GET';
-      const bodyRecord = rawBody ? toBodyRecord(rawBody) : undefined;
+      const bodyRecord =
+        bodyMode === 'raw' && rawJsonBody ? parseRawBody(rawJsonBody) : bodyPairs ? toBodyRecord(bodyPairs) : undefined;
```
[the remark this answers is not in the document]
Comment (SuperFredR) [reply libs/application-generic/src/usecases/preview/preview.usecase.ts]: Fixed in 4b6e451: `preview.usecase.ts` now uses an explicit if/else — when `bodyMode === 'raw'`, only `rawJsonBody` is used (or undefined if empty), no fallback to `bodyPairs`. Same pattern applied in `execute-http-request-step.usecase.ts`.

File: apps/worker/src/app/workflow/usecases/send-message/execute-http-request-step.usecase.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (SuperFredR) [reply apps/worker/src/app/workflow/usecases/send-message/execute-http-request-step.usecase.ts]: Fixed in 4b6e451: `parseRawBody` is now wrapped in try/catch. On failure, an `ExecutionDetails` is created and the step returns FAILED instead of crashing the worker job.

File: apps/dashboard/src/components/workflow-editor/steps/http-request/raw-body-editor.tsx [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+  }
+
+  if (value.includes('{{')) {
+    return undefined;
+  }
+
+  try {
+    const parsed = JSON.parse(value);
+    if (typeof parsed !== 'object' || parsed === null || Array.isArray(parsed)) {
+      return 'Body must be a JSON object';
```
[the remark this answers is not in the document]
Comment (SuperFredR) [reply apps/dashboard/src/components/workflow-editor/steps/http-request/raw-body-editor.tsx]: Fixed in d3373c6: `validateJson` now accepts top-level JSON arrays. The check is just `typeof !== 'object' || === null`.

File: apps/dashboard/src/components/workflow-editor/steps/http-request/raw-body-editor.tsx [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+                isAllowedVariable={isAllowedVariable}
+                variables={variables}
+                onChange={(val) => {
+                  const newVal = typeof val === 'string' ? val : '';
+                  setDraft(newVal);
+                  // Only propagate to the form (and trigger preview update) when valid
+                  if (!validateJson(newVal)) {
+                    field.onChange(newVal);
+                    saveForm();
+                  }
```
[the remark this answers is not in the document]
Comment (SuperFredR) [reply apps/dashboard/src/components/workflow-editor/steps/http-request/raw-body-editor.tsx:81]: Fixed in d3373c6: whitespace-only drafts are now persisted as empty string via `field.onChange(newVal.trim() ? newVal : '')`.

File: apps/worker/src/app/workflow/usecases/send-message/execute-http-request-step.usecase.ts [hunk 11]
```diff
@@ -144,7 +147,36 @@ export class ExecuteHttpRequestStep extends SendMessageType {
...
     const headersRecord = toHeadersRecord(rawHeaders);
-    const bodyObject = toBodyRecord(rawBody);
+
+    let bodyObject: Record<string, unknown> | undefined;
+    try {
+      if (bodyMode === 'raw') {
+        bodyObject = rawJsonBody ? parseRawBody(rawJsonBody) : undefined;
+      } else {
+        bodyObject = toBodyRecord(bodyPairs);
+      }
```
[the remark this answers is not in the document]
Comment (SuperFredR) [reply apps/worker/src/app/workflow/usecases/send-message/execute-http-request-step.usecase.ts]: Fixed in d3373c6: `parseRawBody` now returns `Record<string, unknown> | unknown[]` and accepts arrays. `shouldIncludeBody` and the body object types in execute/preview/test usecases were updated accordingly.

File: apps/dashboard/src/components/workflow-editor/steps/http-request/curl-display.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (SuperFredR) [reply apps/dashboard/src/components/workflow-editor/steps/http-request/curl-display.tsx]: Fixed in 5c19f83: `CurlDisplay` now uses an exclusive raw mode (`if (bodyMode === 'raw') { ... }` with no else fallback to key-value).

File: apps/dashboard/src/components/workflow-editor/steps/http-request/curl-utils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (SuperFredR) [reply apps/dashboard/src/components/workflow-editor/steps/http-request/curl-utils.ts:127]: Fixed in 5c19f83: `buildRawCurlString` is now exclusive in raw mode — no fallback to key-value pairs.

File: apps/dashboard/src/components/workflow-editor/steps/http-request/curl-utils.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (SuperFredR) [reply apps/dashboard/src/components/workflow-editor/steps/http-request/curl-utils.ts]: Good catch — that was a real bug separate from the fallback issue. Fixed in 5881580: added an `escapeShellSingleQuoted` helper that wraps single quotes with `'\\''` and applied it to all interpolated values (URL, headers, signature, raw body, key-value body) in both `curl-utils.ts` and `curl-display.tsx`.

File: apps/dashboard/src/components/workflow-editor/steps/http-request/curl-utils.ts [hunk 12]
```diff
@@ -25,27 +33,42 @@ export function buildRawCurlString(
...
-
-  const bodyStr = bodyObj ? `--data '${JSON.stringify(bodyObj)}'` : '';
-  const parts = [`novu $ curl --location '${url || 'https://api.example.com/endpoint'}'`, headerArgs, bodyStr].filter(
-    Boolean
-  );
+  const parts = [
+    `novu $ curl --location '${escapeShellSingleQuoted(url || 'https://api.example.com/endpoint')}'`,
+    headerArgs,
+    bodyStr,
+  ].filter(Boolean);
```
[the remark this answers is not in the document]
Comment (SuperFredR) [reply apps/dashboard/src/components/workflow-editor/steps/http-request/curl-utils.ts:133]: Fixed in 75cb9c9: `buildRawCurlString` and `CurlDisplay` now include `--request '<METHOD>'` (uppercased and shell-escaped) so the preview always reflects the configured HTTP verb.

**Comment by scopsy** [comment]
@coderabbitai review

**Review by scopsy — APPROVED** [review]

**Comment by scopsy** [comment]
Thank you for the coontribution! I have made a few adjustments on the logic to keep the key-pair and raw json a front end presentation concern and always use the raw json in the saved body
